Add support for CIP42 and EIP1559

aaronmgdr opened the pull request:
### Description

Adds Support for and favors using eip1559 (when no fee currency selected) and cip42 (when it is) legacy will only be used if gasPrice is explicitly set. 

#### why not the new  feecurrency  only tx type?

Because its not available on celo mainnet yet. and was only announced after this was almost done. This PR does however do pre-work vital to getting that in a future release.  unstated Goal with this PR here is to have a version of sdks that are released post the node 18 upgrade.  

### Other changes

#### TopLine Changes

* Add support for CIP42 and EIP1559 Transactions
* Prefer EIP1559 if no feeCurrency and CIP42 if using feeCurrency, Old CeloLegacy transations still supported (set the gasPrice explicitly) but discouraged
* upgrade Web3 dependency

#### @celo/connect

* gasPrice is no longer an option on config. setting 0 (or any empty value) for gasPrice in transaction will result in maxFeePerGas and maxPriorityFee per gas being set on the transaction

* likewise gasPrice is no longer settable on the connection itself

* (note that the gasPrice function for fetching price from node is NOT affected)

* replace connection.fillGasPrice  with connection.setFeeMarketGas

* add inputAccessListFormatter

* remove deprecated setGasPriceForCurrency

* EncodedTransaction Type now supports CIp42 and EIP1559 transactions

* inputCeloTxFormater now returns type FormattedCeloTx although apart from support for cip42 and eip1559 tx the data returned hasnt changed

* numberToHex, ensureLeading0x, inputAddressFormatter now are typed to return `0x{string}` instead of just string

* added parseAccessList and inputAccessListFormatter for converting accessList from json to array of array and back

* RLPEncodedTx now types its transaction field as FormattedTransaction this is more accurate than the CeloTX type it previously had

#### @celo/contrackit

* removed gasPriceSuggestionMultiplier == gas price from rpc no longer multipled by 5

* remove kit.fillGasPrice

* remove updateGasPriceInConnectionLayer

* remove kit.gasPrice

#### @celo/wallet-base

* extractSignature now throws if the length of provided tx is not correct
* extractSignature was incorrectly typed to return Buffers for r and s values

* ensureLeading0x now types output to be `0x{string}`


### protocol tests
* remove underscore dep
* replace custom signing code with LocalWallet.signTransaction

### Tested

see test suite. also made successful tx using celo cli for both new types
### Related issues

- Fixes #10447 
- Fixes #10025 

### Backwards compatibility

No this removes some deprecated methods from contract kit (see changes log above)

### Documentation

aaronmgdr committed:
remove deprecated functions / params for gasPrice / fee currency in kit

aaronmgdr committed:
improve types / format txs

aaronmgdr committed:
signing tests, add viem to compare with in tests.

aaronmgdr committed:
fix type issue, and split the first 12 (not the last 3)

aaronmgdr committed:
remove `only` and give test and actual name

aaronmgdr committed:
remove logs and fix some incorrectly setting tx tyoe to wrong type

aaronmgdr committed:
lint fixes

aaronmgdr committed:
parallelize async calls.

aaronmgdr committed:
cleanup

aaronmgdr committed:
do not extract signer from cip42 txns

this doesnt seem to be a very important feature so punting on it to get CK released

aaronmgdr committed:
lint checks

aaronmgdr committed:
update test

aaronmgdr committed:
revert removal of signer recovery

aaronmgdr committed:
feedback from sep

aaronmgdr committed:
add back signer test for faster deving

aaronmgdr committed:
it works. gandolf rides over the crest as the sun rises blinding the enemy

aaronmgdr committed:
lint

aaronmgdr committed:
show error if the rpc caller  does not attach

aaronmgdr committed:
fix binding

aaronmgdr committed:
lint fix, and remove the call to set FeeMarket (i dont think its actually need as this will happen in tx params normalizer)

aaronmgdr committed:
lint fix

aaronmgdr committed:
no longer doing it like this

aaronmgdr committed:
fix ck tests

aaronmgdr committed:
cheat and pay with CELO to avoid the ganache issue where when paying in fee token querying for gasPrice give and error due to price lookup having an arg and ganache expecting 0 args

aaronmgdr committed:
it was never supposed to be like this

aaronmgdr committed:
why is it gasCurrency here?

aaronmgdr committed:
add some logging for that bug debug -- dont judge me!

aaronmgdr commented on packages/cli/src/base.ts:
```diff
@@ -201,11 +201,9 @@ export abstract class BaseCommand extends Command {
 
     const setStableTokenGas = async (stable: StableToken) => {
       await this.kit.setFeeCurrency(stableTokenInfos[stable].contract)
-      await this.kit.updateGasPriceInConnectionLayer(
-        await this.kit.registry.addressFor(stableTokenInfos[stable].contract)
-      )
```
not needed since eth_gasPrice (feecurrency) now gives price of gas in token

aaronmgdr committed:
logging

aaronmgdr committed:
fix but where  transfer was not adhering to the gasFeeCurrency flag

aaronmgdr committed:
fix #10025 js numbers cause impresission

aaronmgdr committed:
hex must be even

aaronmgdr committed:
cleanup extra logs / commented out code.

add warning for gatewayfee

aaronmgdr committed:
add more logging to try to debug test failure

aaronmgdr committed:
ah why does this exist?

aaronmgdr committed:
whats up here

aaronmgdr committed:
fix test description typos

aaronmgdr committed:
Dont use custom signing code in protocol test

* remove underscore dep

aaronmgdr commented on packages/protocol/package.json:
```diff
@@ -104,7 +104,6 @@
     "@types/mocha": "^7.0.2",
     "@types/targz": "^1.0.0",
     "@types/tmp": "^0.1.0",
-    "@types/underscore": "^1.8.8",
```
weirdly underscore itself was NOT listed as a dep anywhere

montera82 commented on packages/protocol/lib/web3-utils.ts:
```diff
@@ -39,8 +39,8 @@ export async function sendTransactionWithPrivateKey<T>(
   }
-
-  const signedTx: any = await signTransaction(
+  console.info('signing tx with', address)
```
looks like leftover debug code

montera82 commented on packages/protocol/lib/web3-utils.ts:
```diff
@@ -53,7 +53,8 @@ export async function sendTransactionWithPrivateKey<T>(
 
-  const rawTransaction = signedTx.rawTransaction.toString('hex')
+  const rawTransaction = signedTx.raw
+  console.info('sending tx',signedTx.tx, rawTransaction)
```
here as well

montera82 commented on packages/sdk/connect/src/connection.ts:
```diff
@@ -341,20 +326,27 @@ export class Connection {
+    if ((isEmpty(tx.gasPrice) && isEmpty(tx.maxFeePerGas)) || isEmpty(tx.maxPriorityFeePerGas)) {
+      if (isEmpty(tx.maxFeePerGas)) {
+        calls[0] = this.gasPrice(tx.feeCurrency)
+      }
```
this if statements seem interesting, could we make simpler, looks like we are checking tx.maxFeePerGas multiple times?

aaronmgdr commented on packages/sdk/wallets/wallet-base/src/signing-utils.ts:
```diff
@@ -171,44 +391,188 @@ export function extractSignature(rawTx: string): { v: number; r: Buffer; s: Buff
+      const rawValues = RLP.decode(rawTx)
+      debug('signing-utils@recoverTransaction: values are %s', rawValues)
+      const recovery = Bytes.toNumber(rawValues[9])
+      // tslint:disable-next-line:no-bitwise
```
not actually new

montera82 commented on packages/sdk/wallets/wallet-base/src/wallet-base.ts:
```diff
@@ -77,7 +77,8 @@ export abstract class WalletBase<TSigner extends Signer> implements ReadOnlyWall
     const rlpEncoded = rlpEncodedTx(txParams)
-    const addToV = chainIdTransformationForSigning(txParams.chainId!)
+    const addToV =
+      rlpEncoded.type === 'celo-legacy' ? chainIdTransformationForSigning(txParams.chainId!) : 27
```
nit: we could wrap 27 into a const as its kind of magic 🤷

nicolasbrugneaux replied:
+1

montera82 commented on packages/sdk/wallets/wallet-local/src/signing.test.ts:
```diff
@@ -151,44 +39,219 @@ describe('Transaction Utils', () => {
+    describe('Test3 should be able to sign with second account and get the signer back with multiple local accounts', () => {
+      beforeAll(async () => {
+        await setupConnection()
+        // jest.setTimeout(60 * 1000)
```
nit: we can take off the commented code?

aaronmgdr committed:
review  / cleanup

aaronmgdr committed:
protocol now depends on local wallet

aaronmgdr committed:
fix varible lint

montera82 commented on packages/protocol/migrations_ts/28_elect_validators.ts:
```diff
@@ -105,7 +105,11 @@ async function registerValidatorGroup(
+    `    - send funds ${v} to group address ${account.address} from ${privateKeyToAddress(
+      privateKey
+    )}}`
+  )
```
this is needed yea?

aaronmgdr replied:
i guess not

montera82 reviewed (approved):
🚢 it

aaronmgdr committed:
update to 5.0.0-beta.1

aaronmgdr committed:
revert back logging to how it was,
add back ensure 0x to data if it exists
throw an error (like web3 does) if storage key is bad
remove temp file for changes (now in PR description)

nicolasbrugneaux commented on packages/sdk/base/src/address.ts:
```diff
@@ -12,7 +12,8 @@ export const normalizeAddressWith0x = (a: Address) => ensureLeading0x(a).toLower
 
-export const ensureLeading0x = (input: string) => (input.startsWith('0x') ? input : `0x${input}`)
+export const ensureLeading0x = (input: string): `0x${string}` =>
+  input.startsWith('0x') ? (input as `0x${string}`) : (`0x${input}` as const)
```
nit: if we're going to use `0x${string}` as types, maybe alias it as `Address`

nicolasbrugneaux commented on packages/sdk/connect/src/types.ts:
```diff
@@ -9,8 +14,31 @@ export interface CeloParams {
+  gatewayFeeRecipient?: HexOrMissing
+  gatewayFee?: HexOrMissing
+  gas: HexOrMissing
+  gasPrice?: `0x${string}`
```
same comment as above

nicolasbrugneaux commented on packages/sdk/connect/src/types.ts:
```diff
@@ -1,6 +1,11 @@
-
 export type Address = string
 
 export interface CeloParams {
```
could we mark `gatewayFeeRecipient` as deprecated with jsdoc ?

aaronmgdr committed:
Merge branch 'aaronmgdr/new-tx-types' of https://github.com/celo-org/celo-monorepo into aaronmgdr/new-tx-types

aaronmgdr committed:
use named types for address and hex values,
mark gateway fee as deprecated
fix versioning

aaronmgdr committed:
publish sdk 5 and pnp-common 3.0.1

aaronmgdr committed:
add back dev suffix

aaronmgdr committed:
republish -- pnp engine can not be 18

aaronmgdr committed:
add back dev suffix

aaronmgdr committed:
try

aaronmgdr committed:
publish beta

aaronmgdr committed:
Merge branch 'master' into aaronmgdr/new-tx-types

aaronmgdr committed:
publish good release

aaronmgdr committed:
use node 18 in container
